File: leetcode059.py
class Solution:
    def generateMatrix(self, n):
        """
        :type n: int
        :rtype: List[List[int]]
        """

        res = [[0]*n for i in range(n)]
        i = 1
        j = 0
        k = 0
        m = 0
        res[k][m] = i
        while i < n*n:
            if j%4 == 0:

                while m+1 < n and res[k][m+1] ==0 :
                    m = m +1
                    i = i + 1
                    res[k][m] = i

                j=j+1
            if j%4 == 1:

                while k+1 < n and res[k+1][m] == 0:
                    k = k + 1
                    i = i + 1
                    res[k][m] = i

                j = j + 1
            if j % 4 == 2:

                while m-1>=0 and res[k][m -1] == 0:
                    m = m - 1
                    i = i + 1
                    res[k][m] = i

                j = j + 1
            if j % 4 == 3:

                while k-1>=0 and res[k-1][m] == 0:
                    k = k -1
                    i = i + 1
                    res[k][m] = i

                j = j + 1
        return res

if __name__ == "__main__":
    s = Solution()
    print(s.generateMatrix(3))

File: leetcodec++/leetcode059.cpp
//
// Created by 黄敏辉 on 2019-07-17.
//

class Solution {
public:
    vector<vector<int>> generateMatrix(int n) {
        vector<vector<int>> res;
        if(n<=0) return res;
        res.resize(n, vector<int>(n,0));

        int lr = 0, lc = -1;
        int dir = 0;
        int val = 1;
        vector<vector<int>> dirs = {{0,1}, {1,0}, {0,-1}, {-1, 0}};

        vector<int> nSteps = {n, n - 1};

        while(nSteps[dir%2]){

            for(int i = 0; i < nSteps[dir%2]; i++){
                lr += dirs[dir][0];
                lc += dirs[dir][1];
                res[lr][lc] = val;
                val += 1;
            }

            nSteps[dir%2]--;
            dir = (dir + 1)%4;
        }
        return res;
    }
};